middle_end/flambda2/types/expand_head.ml

```diff
@@ -389,56 +391,114 @@ let make_suitable_for_environment env (to_erase : to_erase) bind_to_and_types =
+            not (TE.mem suitable_for (Name.var var))
+          | All_variables_except to_keep -> not (Variable.Set.mem var to_keep)
+        in
+        Name_occurrences.fold_variables free_vars ~init:([], [], [])
+          ~f:(fun
+               (( unavailable_vars_renamed,
+                  unavailable_vars_expanded,
+                  unavailable_vars_removed ) as unavailable_vars)
+               var
+             ->
+            if erase var
+            then
+              if Name_occurrences.mem_var free_vars_except_through_closure_vars
+                   var
+              then
+                match Name_occurrences.count_variable free_vars var with
+                | Zero ->
+                  Misc.fatal_errorf
+                    "Inconsistent occurrences of %a in free names"
+                    Variable.print var
+                | One ->
+                  ( unavailable_vars_renamed,
+                    var :: unavailable_vars_expanded,
+                    unavailable_vars_removed )
+                | More_than_one ->
+                  ( var :: unavailable_vars_renamed,
+                    unavailable_vars_expanded,
+                    unavailable_vars_removed )
+              else
+                ( unavailable_vars_renamed,
+                  unavailable_vars_expanded,
+                  var :: unavailable_vars_removed )
+            else unavailable_vars)
       in
       (* Fetch the type equation for each free variable. Also add in the
          equations about the "bind-to" names provided to this function. If any
          of the "bind-to" names are already defined in [env], the type given in
-         [bind_to_and_types] takes precedence over such definition. *)
+         [bind_to_and_types] takes precedence over such definition. All
+         occurrences of variables that only occur once are expanded directly.
+         All occurrences of variables that are only reachable through closure
+         variables are replaced with an Unknown type. *)
+      let to_expand = Variable.Set.of_list unavailable_vars_expanded in
+      let to_remove = Variable.Set.of_list unavailable_vars_removed in
+      let to_project = Variable.Set.union to_expand to_remove in
+      let expand_type ty =
+        let rec expand var =
+          let ty = TE.find env (Name.var var) None in
+          if Variable.Set.mem var to_remove
+          then MTC.unknown_like ty
+          else
+            match TG.get_alias_exn ty with
+            | exception Not_found ->
+              TG.project_variables_out ~to_project ~expand ty
+            | simple ->
+              Simple.pattern_match' simple
+                ~const:(fun _ -> ty)
+                ~symbol:(fun _ ~coercion:_ -> ty)
+                ~var:(fun var ~coercion ->
+                  if Variable.Set.mem var to_expand
+                  then TG.apply_coercion (expand var) coercion
+                  else ty)
+        in
+        TG.project_variables_out ~to_project ~expand ty
```

Comment:
The key pain point is that it's not possible to refer to a typing environment inside `Type_grammar.ml`. It might be possible to only pass a `find_type` function instead of the `expand` function though.